controllers/toolchaincluster/toolchaincluster_controller_test.go

```diff
@@ -101,23 +103,81 @@ func TestClusterControllerChecks(t *testing.T) {
 		assertClusterStatus(t, cl, "stable", healthy())
 	})
 
-	t.Run("toolchain cluster cache not found", func(t *testing.T) {
+	t.Run("Checking the run check health default ", func(t *testing.T) {
```

Comment:
I feel like this should not be tested here as it is testing the Healthchecker code.
Indeed, if the `CheckHealth` field is not set, we are testing the Healthcheker code that's executed at `controllers/toolchaincluster/toolchaincluster_controller.go#L86`.

To me in this suite we should just test that (1) if we got an error we are forwarding the error and (2) if we got a set of conditions we are setting the condition in the status.

Another thing that we'd be interested in testing is "if CheckHealth is not tested, we are invoking the healthchecker", but it's seems to me is a quite hard to test in unit tests and not being that useful. E2E can easily cover this scenario.

Reply:
i think CheckHealth is a part of controller and checking it from somewhere else wouldn't make sense i think

Reply:
yes, but for how the reconciler is designed if its `CheckHealth` field is not set, it would use the Healthchecker logic. This scenario inadvertently checks the Healthchecker logic.  

To verify it, you can 
```bash
go test ./controllers/toolchaincluster/... -run 'TestClusterControllerChecks/Checking_the_run_check_health_default_' -coverprofile cover.out
go tool cover -html=cover.out
```

the output should be similar to the following:
```diff
+func (r *Reconciler) runCheckHealthOrDefault(ctx context.Context, rcc *kubeclientset.Clientset) []toolchainv1alpha1.Condition {
+        if r.CheckHealth != nil {
-                return r.CheckHealth(ctx, rcc)
+        }
+        hcond := getClusterHealthStatus(ctx, rcc)
+        return hcond
+}
```

this means that we are invoking -hence testing- the healthchecker code. 

To me the healthchecker logic should be tested in isolation, in its own unit tests. Here we should just test that the reconciler handles correctly the different kind of result the func `r.runCheckHealthOrDefault(ctx, clientSet)` can return, i.e. empty set of conditions, conditions already in status, conditions not in status, etc.

To avoid testing the healthchecker code in the reconciler tests we should enforce the `CheckHealth` field to be always set.

Reply:
updated it